src/v/utils/tracking_allocator.h

```diff
@@ -0,0 +1,169 @@
+    mem_tracker(mem_tracker&) = delete;
+    mem_tracker(mem_tracker&&) = default;
+    mem_tracker& operator=(mem_tracker&) = delete;
+    mem_tracker& operator=(mem_tracker&&) = delete;
+    ~mem_tracker() = default;
+
+    void allocate(int64_t bytes) { _consumption += bytes; }
+
+    void deallocate(int64_t bytes) { _consumption -= bytes; }
+
+    ss::lw_shared_ptr<mem_tracker> create_child(ss::sstring);
+
+    /// Consumption of self + all the children rooted with this tracker..
+    int64_t consumption() const;
+
+    /// Pretty prints the tracker tree rooted at this instance.
+    /// Example:
+    /// |
+    /// root: 32.000bytes
+    /// | |
+    /// | child1: 24.000bytes
+    /// | | |
+    /// | | child11: 0.000bytes
+    /// | | | |
+    /// | | | child111: 0.000bytes
+    /// | |
+    /// | child2: 0.000bytes
+    ss::sstring debug(int level = 0) const;
+
+private:
+    ss::sstring _label;
+    int64_t _consumption = 0;
+    intrusive_list_hook _hook;
+    intrusive_list<mem_tracker, &mem_tracker::_hook> _children;
+};
+
+/// An allocator paired with a memory tracker to account for consumption.
+template<class T, class allocator = std::allocator<T>>
+class tracking_allocator : public allocator {
+public:
+    explicit tracking_allocator(ss::lw_shared_ptr<mem_tracker> tracker)
+      : _tracker(tracker) {}
+
+    template<class U>
+    explicit tracking_allocator(const tracking_allocator<U>& other)
+      : allocator(other)
+      , _tracker(other.tracker()) {}
+
+    template<class U>
+    struct rebind {
+        using other = tracking_allocator<U, std::allocator<U>>;
+    };
+
+    [[nodiscard]] T* allocate(std::size_t n) {
+        _tracker->allocate(n * sizeof(T));
+        return allocator::allocate(n);
+    }
+
+    void deallocate(T* p, std::size_t n) {
+        allocator::deallocate(p, n);
+        _tracker->deallocate(n * sizeof(T));
+    }
+
+    ss::lw_shared_ptr<mem_tracker> tracker() const { return _tracker; }
+
+private:
+    ss::lw_shared_ptr<mem_tracker> _tracker;
+};
+
+namespace mem_tracked {
+/// Since allocator is typically the last parameter for stl containers,
+/// these template utilities help avoid template boiler plate by auto
+/// setting the intermediate parameters to defaults.
+template<class T>
+concept is_map = requires {
+    typename T::key_type;
+    typename T::value_type;
+    typename T::mapped_type;
+    typename T::hasher;
+    typename T::key_equal;
+    typename T::allocator_type;
+};
+
+template<class T>
+concept is_set = requires {
+    typename T::key_type;
+    typename T::value_type;
+    typename T::key_compare;
+    typename T::value_compare;
+    typename T::allocator_type;
+};
+
+template<
+  template<class...>
+  class Map,
+  class K,
+  class V,
+  class Hasher = typename Map<K, V>::hasher,
+  class KeyEq = typename Map<K, V>::key_equal,
+  class Allocator = tracking_allocator<std::pair<const K, V>>>
+using map_t = Map<K, V, Hasher, KeyEq, Allocator>;
```

Comment:
actually I had tracked_map_.. in the first place but mem_tracked::tracked_map.. seemed a bit too much, wdyt.

Reply:
I'm fine with the public API being `mem_tracked::map()`, but for these helpers_t I thought it might be helpful to distinguish them (we're juggling a bunch of "map" and "set" word variants here). I don't feel strongly about it either way.

Reply:
Ah actually I misconstrued these as helpers when they're actually the resulting tracked types. I'm okay with this.